Example/tvOSSample/tvOSSample/StorageViewController.swift

```diff
@@ -0,0 +1,148 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import UIKit
+import FirebaseStorage
+
+class StorageViewController: UIViewController {
+  /// An enum describing the different states of the view controller.
+  private enum UIState: Equatable {
+    /// No image is being shown, waiting on user action.
+    case cleared
+
+    /// Currently downloading from Firebase.
+    case downloading(StorageTask)
+
+    /// The image has downloaded and should be displayed.
+    case downloaded(UIImage)
+
+    /// Show an error message and stop downloading.
+    case failed(String)
+
+    /// Equatable support for UIState.
+    static func ==(lhs: StorageViewController.UIState, rhs: StorageViewController.UIState) -> Bool {
+      switch (lhs, rhs) {
+      case (.cleared, .cleared): return true
+      case (.downloading, .downloading): return true
+      case (.downloaded, .downloaded): return true
+      case (.failed, .failed): return true
+      default: return false
+      }
+    }
+  }
+
+  /// MARK: - Properties
+
+  /// The current internal state of the view controller.
+  private var state: UIState = .cleared {
+    didSet { changeState(from: oldValue, to: state) }
+  }
+
+  // MARK: Interface
+
+  /// Image view to display the downloaded image.
+  @IBOutlet weak var imageView: UIImageView!
+
+  /// The download button.
+  @IBOutlet weak var downloadButton: UIButton!
+
+  /// The clear button.
+  @IBOutlet weak var clearButton: UIButton!
+
+  /// A visual representation of the state.
+  @IBOutlet weak var stateLabel: UILabel!
+
+  // MARK: - User Actions
+
+  @IBAction func downloadButtonHit(_ sender: UIButton) {
+    guard case .cleared = state else { return }
+
+    // Start the download!
+    let storage = Storage.storage()
+    let ref = storage.reference(withPath: Constants.downloadPath)
+    let task = ref.getData(maxSize: Constants.maxSize) { [unowned self] (data, error) in
+      guard let data = data else {
+        self.state = .failed("Error downloading: \(error!.localizedDescription)")
+        return
+      }
+
+      // Create a UIImage from the PNG data.
+      guard let image = UIImage(data: data) else {
+        self.state = .failed("Unable to initialize image with data downloaded.")
+        return
+      }
+
+      self.state = .downloaded(image)
+    }
+
+    // We could have the image cached already causing the completion block to run before this line.
+    // If that's the case, we don't need to do anything else and can return.
+    if case .downloaded = state { return }
```

Comment:
Firebase Storage doesn't have a cache. Can you explain what is going on here? Thanks!

Reply:
Huh, interesting. There were a few situations where if you already downloaded the photo + cleared it, the completion block setting the state to `downloaded` would be triggered *before* the code afterward setting it to `downloading`. Maybe this is Apple's caching under the hood?

Reply:
It could be using an NSURLCache by default, but even with that I would be surprised that this code would race. There might be something else going on here, but I am not quite sure what it could be.

Reply:
Updated, thanks!